[addons] Regarding addon.xml "<news>", use changelog.txt as an alternative

AlwinEsch opened the pull request:
## Description

In the event that the changes log (news) in addon.xml is empty, check whether it is an installed addon and read a changelog.txt as a replacement, if available.

Many addons use this "changelog.txt", but less use the `<news>` field in addon.xml, and nobody actually uses the translation path in it (once unknown and difficult to keep correct in the case of continuous changes).
Hence, this brings a fall back to reading the text file as a replacement.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

Here about this file https://github.com/xbmc/audiodecoder.sacd/blob/Matrix/audiodecoder.sacd/changelog.txt given to it:
![Bildschirmfoto von 2021-09-30 14-39-01](https://user-images.githubusercontent.com/6879739/135457690-f9491d8e-c958-4c15-b9a3-72214083c21b.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented:
Definitely +1. I never liked to pollute addons.xml with `<news>`.

AlwinEsch commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -354,8 +355,26 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
+       * whether it is an installed addon and read a changelog.txt as a
+       * replacement, if available. */
       GetTextList(child, "news", addon->m_changelog);
+      if (addon->m_changelog.empty() && repo.artdir.empty() && !addonPath.empty())
```
This field is only used for repos, so if something is there, changelog.txt can never be present.
If this were not possible, it would try to find the "changelog.txt" in the network for every addon

AlwinEsch replied:
Added a new commit, which will set it to a new value. 
I don't want to do it as an extra function variable because `const CRepository::DirInfo& repo` is only set if it comes from the repo anyway.

AlwinEsch commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -13,6 +13,7 @@
 #include "addons/addoninfo/AddonType.h"
 #include "filesystem/File.h"
 #include "filesystem/SpecialProtocol.h"
+#include "utils/auto_buffer.h"
```
Ah crap, forget the stupid upper / lowercase, must go further down.

phunkyfish commented:
I love this change. Then we can stop populating both news and changelog

phunkyfish commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -117,6 +117,8 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
     return false;
   }
 
+  const bool isRepoXMLContent = !repo.artdir.empty() && !repo.datadir.empty();
```
This should have a comment explaining why this means it’s repo xml content.

AlwinEsch replied:
Added comment

AlwinEsch committed:
[addons] use new value isRepoXMLContent on CAddonInfoBuilder::ParseXML

The background to this is the need to distinguish between the locally installed add-on and the repo-stored add-on, so that the currently used type can be better understood.
Previously this was only possible with `repo.artdir.empty()` which made it difficult to see the background.

AlwinEsch commented:
Jenkins build this please

AlwinEsch commented:
Request OK to come in?

phunkyfish reviewed (approved):
Yes, fine by me.

AlwinEsch commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -354,8 +361,26 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
+        if (CFile::Exists(changelog))
+        {
+          CFile file;
+          auto_buffer buf;
```
Have changed now to `using XFILE::auto_buffer;`, seen his place before on `XUTILS` and why used first.
You has mean this by the comment?

Reason about the `using XFILE::...`, find it good to reduce line length and to know in related area, to where relates it, by a `using namespace XFILE` on top goes it a bit away to know where a part relates.

phunkyfish replied:
No, my comment was about using `auto_buffer` instead if `autoBuffer`.

AlwinEsch replied:
Aha, now I understand. Seen his class naming here https://github.com/xbmc/xbmc/blob/master/xbmc/filesystem/File.h#L73 and relates to https://github.com/xbmc/xbmc/blob/master/xbmc/utils/auto_buffer.h, that's why I not come direct to it, as used everywhere with it.

Can add a `using CAutoBuffer = XFILE::auto_buffer;` to have as `CAutoBuffer there`?

AlwinEsch replied:
Think is from the old days, when the guidelines were not used that much.

phunkyfish replied:
Ah, ok. Then leave it as it is. Good to go in.

AlwinEsch committed:
[addons] Regarding addon.xml "<news>", use changelog.txt as an alternative

In the event that the changes log (news) in addon.xml is empty, check whether it is an installed addon and read a changelog.txt as a replacement, if available.

Many addons use this "changelog.txt", but less use the "<news>" field in addon.xml, and nobody actually uses the translation path in it (once unknown and difficult to keep correct in the case of continuous changes).
Hence, this brings a fall back to reading the text file as a replacement.

phunkyfish commented:
jenkins build this please

phunkyfish commented:
I rebased this PR to clear up the IOS/TV errors. Not sure what the Android failure are from. Maybe it will clear those up to.
